Explicitly forbidden the other inherited methods of RemoteModule.

wayi1 opened the pull request:
Summary:
Add comments on the other inherited methods of RemoteModule.

In addition, L135 was reformatted by lint.

Test Plan:
N/A, as only comments are added.

Couldn't find any calls or any documentations of the other RemoteModule's methods that are inherit from nn.Module. Therefore, played with fbsource/fbcode/caffe2/torch/testing/_internal/distributed/nn/api/remote_module_test.py to validate some clarifications here.
buck test test/distributed/rpc:process_group_agent -- RemoteModule

Differential Revision: D23392842

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -50,6 +57,16 @@ def __init__(
 
+        Since all the other methods except ``forward`` and ``forward_async``
+        directly inherit from ``nn.Module``,
+        these methods are only applied to the caller on the local side,
```
Maybe reword this a bit, without reading the following sentence it seems like these methods actually work for RemoteModule and are applied on the local side. Maybe we can just mention that apart from `forward` and `forward_async` no other methods are supported from nn.Module.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -50,6 +57,16 @@ def __init__(
+        these methods are only applied to the caller on the local side,
+        instead of the user-provided ``module_cls`` that will be executed on the remote node.
+        Therefore, normally these methods should not be used, and the usage will throw exceptions.
+        Particularly, to create a hybrid model, typically the local modules should be
+        in parallel with remote modules, rather than as submodules of any remote module.
```
I'm not sure I followed this comment completely, what do you mean by the modules should be in "parallel" to remote modules?

SciPioneer replied:
I mean the relationship of any local modules with a remote module in a hybrid model should be the peers, rather than containment by calling add_module method.

Example:
class HybridModel(nn.Module):
    def __init__(self):
        nn.Module.__init__(self)
        self.remote_embedding = RemoteModule(
            nn.Embedding,
            num_embeddings=1024,
            embedding_dim=4,
            on='embedding_server')
        self.local_linear = nn.Linear(
            in_features=4,
            out_features=2,
        )

Updated the comment.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -50,6 +57,16 @@ def __init__(
+        Particularly, to create a hybrid model, typically the local modules should be
+        in parallel with remote modules, rather than as submodules of any remote module.
+        See the hybrid model example:
+        fbsource/fbcode/caffe2/torch/fb/distributed/model_parallel/docs/README.md
```
This is an internal file (anything with fb in the path is internal only), let's not reference internal only code in OSS.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -143,6 +158,114 @@ def __init__(
             setattr(self, method_name, types.MethodType(method, self))
 
+    def _raise_not_supported(self):
+        raise ValueError("Method not supported for RemoteModule")
```
nit: Is it possible to add the method name to this error?

SciPioneer replied:
Sure. Passed method.__name__ to this function.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -143,6 +158,114 @@ def __init__(
+        _raise_not_supported()
+
+    def parameters(self, recurse: bool = True) -> Iterator[Parameter]:
+        _raise_not_supported()
```
Maybe this one should mention that we should use `remote_parameters()` instead in the error message.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -50,6 +57,17 @@ def __init__(
+        Particularly, to create a hybrid model, typically the local modules should be
+        created outside of remote modules, rather than as submodules of any remote module (by calling ``add_module``).
+        Hybrid Example:
+        class HybridModel(nn.Module):
+            def __init__(self):
+                nn.Module.__init__(self)
+                self.remote_embedding = RemoteModule(...)
+                self.local_linear = nn.Linear(...)
+
```
nit: You probably want to put this into a codeblock (ex: https://github.com/pytorch/pytorch/blob/master/torch/distributed/rpc/api.py#L257) so that it shows up in a code block appropriately in our docs.

SciPioneer replied:
Great advice! Thx!

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -143,6 +159,114 @@ def __init__(
             method = torch.jit.export(method)
             setattr(self, method_name, types.MethodType(method, self))
 
+    def _raise_not_supported(self, name):
+        raise ValueError("Method ``{}`` not supported for RemoteModule".format(name))
```
Let's add a simple unit in `remote_module_test.py` that validates (using self.assertRaisesRegex) all of these throw appropriate exceptions.

Commit:
Explicitly forbidden the other inherited methods of RemoteModule. (#43895)

Summary:
Throw exceptions when the methods except for forwardXXX are used.

Original PR issue: RemoteModule enhancements #40550

Pull Request resolved: https://github.com/pytorch/pytorch/pull/43895

Test Plan: buck test test/distributed/rpc:process_group_agent -- RemoteModule

Reviewed By: rohan-varma

Differential Revision: D23392842

Pulled By: SciPioneer

fbshipit-source-id: 1d837afbf333cb667574310d9a55f39a4aa93c57
